examples/energy-management-app/energy-management-common/src/ChargingTargetsMemMgr.cpp

```diff
@@ -0,0 +1,191 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app-common/zap-generated/cluster-objects.h>
+
+#include "ChargingTargetsMemMgr.h"
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::DataModel;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::EnergyEvse;
+
+ChargingTargetsMemMgr::ChargingTargetsMemMgr() : mChargingTargetSchedulesIdx(0), mNumDailyChargingTargets(0)
+{
+    memset(mpListOfDays, 0, sizeof(mpListOfDays));
+}
+
+ChargingTargetsMemMgr::~ChargingTargetsMemMgr()
+{
+    // Free all memory allocated for the charging targets
+    for (uint16_t idx = 0; idx < kEvseTargetsMaxNumberOfDays; idx++)
+    {
+        if (mpListOfDays[idx] != nullptr)
+        {
+            chip::Platform::Delete(mpListOfDays[idx]);
+        }
+    }
+}
+
+void ChargingTargetsMemMgr::Reset(uint16_t chargingTargetSchedulesIdx)
+{
+    // MUST be called for each entry in DataModel::List<const Structs::ChargingTargetScheduleStruct::Type> chargingTargetSchedules
+    mNumDailyChargingTargets    = 0;
+    mChargingTargetSchedulesIdx = chargingTargetSchedulesIdx;
+
+    // Free up any memory associated with this targetSchedule
+    if (mpListOfDays[mChargingTargetSchedulesIdx] != nullptr)
+    {
+        chip::Platform::MemoryFree(mpListOfDays[mChargingTargetSchedulesIdx]);
+        mpListOfDays[mChargingTargetSchedulesIdx] = nullptr;
+    }
+}
+
+void ChargingTargetsMemMgr::AddChargingTarget(const EnergyEvse::Structs::ChargingTargetStruct::Type & chargingTarget)
+{
+    if (mNumDailyChargingTargets < kEvseTargetsMaxTargetsPerDay)
+    {
+        mDailyChargingTargets[mNumDailyChargingTargets++] = chargingTarget;
+    }
+}
+
+EnergyEvse::Structs::ChargingTargetStruct::Type * ChargingTargetsMemMgr::GetChargingTargets() const
+{
+    return mpListOfDays[mChargingTargetSchedulesIdx];
+}
+
+uint16_t ChargingTargetsMemMgr::GetNumDailyChargingTargets() const
+{
+    return mNumDailyChargingTargets;
+}
+
+CHIP_ERROR ChargingTargetsMemMgr::AllocAndCopy()
+{
+    // NOTE: ChargingTargetsMemMgr::Reset() must be called as specified in the class comments in
+    // ChargingTargetsMemMgr.h before this method can be called.
+
+    if (mNumDailyChargingTargets > 0)
+    {
+        // Allocate the memory first and then use placement new to initialise the memory of each element in the array
+        mpListOfDays[mChargingTargetSchedulesIdx] = static_cast<EnergyEvse::Structs::ChargingTargetStruct::Type *>(
+            chip::Platform::MemoryAlloc(sizeof(EnergyEvse::Structs::ChargingTargetStruct::Type) * mNumDailyChargingTargets));
+
+        if (mpListOfDays[mChargingTargetSchedulesIdx] == nullptr)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
```

Comment:
```suggestion
        VerifyOrReturnError(mpListOfDays[mChargingTargetSchedulesIdx] != nullptr, CHIP_ERROR_NO_MEMORY);
```

is slightly more idiomatic in chip codebase